network/p2p/gossip/gossip.go

```diff
@@ -180,6 +190,112 @@ func (p *PullGossiper[T, U]) handleResponse(
 	p.receivedBytes.Add(float64(receivedBytes))
 }
 
+// NewPushGossiper returns an instance of PushGossiper
+func NewPushGossiper[T Gossipable](client *p2p.Client, namespace string) *PushGossiper[T] {
+	return &PushGossiper[T]{
+		client: client,
+		sentN: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "gossip_sent_n",
+			Help:      "amount of gossip sent (n)",
+		}),
+		sentBytes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "gossip_sent_bytes",
+			Help:      "amount of gossip sent (bytes)",
+		}),
+	}
+}
+
+// PushGossiper broadcasts gossip to peers randomly in the network
+type PushGossiper[T Gossipable] struct {
+	client *p2p.Client
+
+	sentN     prometheus.Counter
+	sentBytes prometheus.Counter
+
+	lock   sync.Mutex
+	queued []T
+}
+
+// Gossip flushes any queued gossipables
+func (p *PushGossiper[T]) Gossip(ctx context.Context) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if len(p.queued) == 0 {
+		return nil
+	}
+
+	msg := &sdk.PushGossip{
+		Gossip: make([][]byte, 0, len(p.queued)),
+	}
+
+	sentBytes := 0
+	for _, tx := range p.queued {
+		bytes, err := tx.Marshal()
+		if err != nil {
+			return err
+		}
+
+		msg.Gossip = append(msg.Gossip, bytes)
+		sentBytes += len(bytes)
```

Comment:
I wonder if this should be limited.

Reply:
We probably should... if we don't we could end up in a world where no txs are being gossiped if a node is always having a huge mempool since they can breach the max message size and just keep getting dropped. We can pass in a max size into the constructor to guarantee progress here.